frontend/lib/framework/resolution-error-classes-list.cpp

```diff
@@ -811,24 +811,131 @@ void ErrorUnstable::write(ErrorWriterBase& wr) const {
   return;
 }
 
+// describe where a symbol came from
+// (in a way, it prints out a ResultVisibilityTrace)
+// 'start' indicates where in the trace to start, since sometimes
+// the first element might have already been printed.
+// 'oneOnly' indicates that only the 1st match should be described
+//
+// if 'intro' will be emitted before the first message for a trace
+// (only relevant if start==0).
+static void describeSymbolSource(ErrorWriterBase& wr,
+                                 const uast::AstNode* ast,
+                                 UniqueString name,
+                                 const resolution::BorrowedIdsWithName& match,
+                                 const resolution::ResultVisibilityTrace& trace,
+                                 int start,
+                                 bool oneOnly,
+                                 const char* intro) {
+  CHPL_ASSERT(0 <= start);
+
+  bool encounteredAutoModule = false;
+  UniqueString from = name;
+  int n = trace.visibleThrough.size();
+  bool first = true;
+  for (int i = start; i < n; i++) {
+    const auto& elt = trace.visibleThrough[i];
+    if (elt.automaticModule) {
+      std::string msg;
+      if (start==0 && first) {
+        msg = intro;
+      }
+      if (from != name) {
+        msg += "'" + from.str() + "'";
+      } else {
+        msg += "it";
+      }
+      wr.note(ast, msg, " was provided by the automatically-included modules");
+      encounteredAutoModule = true;
+      first = false;
+      break;
+    } else if (elt.fromUseImport) {
+      std::string errbegin;
+      std::string nameSuffix;
+      if (start==0 && first) {
+        errbegin = intro;
+        errbegin += "through";
+      } else {
+        errbegin = "and then through";
+      }
+      if (from != name) {
+        nameSuffix += " providing '" + from.str() + "'";
+      }
+
+      wr.note(locationOnly(elt.visibilityClauseId), errbegin, " the '", elt.visibilityStmtKind, "' statement", nameSuffix, " here:");
+      wr.code<ID,ID>(elt.visibilityClauseId, { elt.visibilityClauseId });
+      from = elt.renameFrom;
+      first = false;
```

Comment:
It's intended to literally reflect whether or not we are writing the first line of output, because if it is the first line we want to include `intro`. I'll add a comment to its declaration, though.